### PR TITLE
TT-3648 started to process oauth clients events

### PR DESCRIPTION
**Pull request by sredxny**
## Description
Now gateway process any signal of oauth client event from mdcb, it will process any: create, update and delete operation. It will allow slaves datacenters to store locally the oauth clients so it will not hit mdcb everytime that an operation related to auth is performed. 

## Related Issue
https://tyktech.atlassian.net/browse/TT-3648

## Motivation and Context
Make slave datacenter more resilient to mdcb outages

## How This Has Been Tested
- Ran the whole mdcb environment. Using the branches of sink and dashboard: TT-3646-feature/emit-oauth-events and 3647/feature-process-oauth-events
- Create an api with ouath enabled via dashboard
- via dashboard created an oauth client for that api and after some seconds we should see that client in the slave redis
- via dashboard update the oauth client and after some seconds the changes should be visible in the slave redis
- via dashboard delete the oauth client and after some seconds we should not see the oauth client in the slave redis neither the register in the index key for oauth clients
- Created client
- Shut down sink
- generated access token in worker node

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [x] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

### CODE REVIEW

File: gateway/rpc_storage_handler.go [hunk 1]
```diff
@@ -817,13 +820,67 @@ func (gw *Gateway) getSessionAndCreate(keyName string, r *RPCStorageHandler, isH
 	}
 }
 
+func (gw *Gateway) ProcessOauthClientsOps(clients map[string]string) {
```
Comment (tbuchaillot) [inline gateway/rpc_storage_handler.go:823]: It looks good! can we add a func description?

File: tyk-gateway.pid [hunk 2]
```diff
@@ -1 +1 @@
-43267
\ No newline at end of file
+69425
```
Comment (tbuchaillot) [inline tyk-gateway.pid:1]: I think we should discard this

**Review by tbuchaillot — APPROVED** [review]
Looks good! it would be great if we add some code comments 👍

File: tyk-gateway.pid (hunk 2 above)
Comment (sredxny) [reply tyk-gateway.pid]: I added this file into the gitignore file, it was introduced by mdcb testing framework and we dont need to keep track of changes on this file

File: gateway/rpc_storage_handler.go (hunk 1 above)
Comment (sredxny) [reply gateway/rpc_storage_handler.go:825]: done

File: gateway/rpc_storage_handler.go [hunk 3]
```diff
@@ -817,13 +820,69 @@ func (gw *Gateway) getSessionAndCreate(keyName string, r *RPCStorageHandler, isH
...
+		if err != nil {
+			log.Error("Could not get oath storage for api")
+			return
+		}
+		switch action {
+		case OauthClientAdded:
+			// on add: pull from rpc and save it in local redis
+			client, err := store.GetClient(oauthClientId)
+			if err != nil {
+				log.WithError(err).Error("Could not retrieve new oauth client information")
```
Comment (matiasinsaurralde) [inline gateway/rpc_storage_handler.go:843]: Should we `return` after this fails or the expected behavior is to continue?

File: gateway/rpc_storage_handler.go [hunk 4]
```diff
@@ -817,13 +820,69 @@ func (gw *Gateway) getSessionAndCreate(keyName string, r *RPCStorageHandler, isH
...
+				log.WithError(err).Error("Could not retrieve oauth client information")
+			}
+			err = store.DeleteClient(oauthClientId, orgID, false)
+
+			if err != nil {
+				log.WithError(err).Error("Could not delete oauth client")
+			}
+			client, err = store.GetClient(oauthClientId)
+			if err != nil {
+				log.WithError(err).Error("Could not retrieve oauth client information")
```
Comment (matiasinsaurralde) [inline gateway/rpc_storage_handler.go:868]: Same question as the previous one, for all the above conditions.

**Review by matiasinsaurralde — APPROVED** [review]
Looks good, have added a few comments

File: gateway/rpc_storage_handler.go (hunk 3 above)
Comment (sredxny) [reply gateway/rpc_storage_handler.go:843]: Thank you for pointing this out, indeed it should be a "continue" statement, so other events are processed as well. I just uploaded the change

**Pull request merged**